### ISSUE TITLE
[WT-2] Training scripts and hyper-parameters

### ISSUE THREAD

**Issue by DavidHerel** [KhaleelKhan/EvNN#4]
Hi there,

could you please provide training scripts and hyper-parameters to obtain the same results as in the paper for WT-2 dataset?
Or you can provide already pre-trained models.

Thank you,
David

**Comment by markschoene** [comment]
Hi @DavidHerel, thanks for your interest in our work! We are happy to provide training scripts and checkpoints. Please note that we implemented some improvements after publishing at ICLR, which I just pushed to this repository (see #5 ).
- Use AdamW to optimize the models
- Simplify the parameterization of the thresholds
- Changed the initialization of the thresholds

Some of the improvements that we observe:
- faster convergence
- more balanced activity

To train EGRU with 1350 hidden units on Wikitext-2, use
If you want to train EGRU, I would suggest to switch to this updated branch at 21348c100a1dd3a1dd039329b91abb4109b241a7, and run
```python
python benchmarks/lm/train.py --data=/your/data/directory --scratch=/your/scratch/directory/Experiments --dataset=WT2 --epochs=800 --batch_size=128 --rnn_type=egru --layer=3 --bptt=70 --scheduler=cosine --weight_decay=0.12 --learning_rate=0.001 --learning_rate_thresholds 0.0 --emb_dim=750 --dropout_emb=0.7 --dropout_words=0.1 --dropout_forward=0.25 --grad_clip=0.1 --thr_init_mean=0.01 --dropout_connect=0.7 --hidden_dim=1350 --pseudo_derivative_width=3.6 --scheduler_start=400 --seed=913420
```
On our machine this gets test perplexity of 68.4 on wikitext-2 with about 25% active neurons (75% sparsity) and 60% backward sparsity.

Checkpoints for EGRU with 1350 hidden units can be [downloaded from this link](https://cloudstore.zih.tu-dresden.de/index.php/s/NPQ9pLnpZnTsM5X).

We also observe that EGRU benefits from larger hidden states on wikitext-2.  I would have to retrain these models though with the updated repository if you are interested in checkpoints. Please get in touch if you require larger models.

**Comment by DavidHerel** [comment]
Thank you so much for quick response!
I would love to try bigger models, however EGRU with 1350 units should be fine for my purposes as it is same as in paper. 

One last thing. If I would like to evaluate this model on dev and test set. I guess I should use eval.py script with --load argument?

**Comment by markschoene** [comment]
I updated the inference script, and the benchmarks/lm/README.md on the branch referenced above. The inference script also requires a config file to build the model before loading the checkpoint. I will upload this once I have high bandwidth internet again on Monday.

In the meantime, you could use the train.py script with `--checkpoint` argument to load the checkpoint and train for 1 epoch ;)

**Comment by markschoene** [comment]
Updated the download link. It now contains checkpoints and config files. To run inference, please download the checkpoints and configs and place them in the same folder structure as listed in the download link. Then follow the instructions at benchmarks/lm/README.md
```
WT2/
├─ checkpoints/
│  ├─ EGRU_best_model.cpt
├─ config.yaml
```
We are happy to help if there are any problems.

**Comment by DavidHerel** [comment]
Thanks so much for the help.
I have one more question. With the linked models and the code I am able to obtain:

61.21 ppl on test set on PTB and 69.40 on WT2.
In the paper the results are 57.0 on PTB and 68.9 on WT2.

WT2 is almost identical as in the paper, so that is great. But do you think that you could also upload a model on PTB that is more close to the reported ppl?

Thanks so much.
David

**Comment by DavidHerel** [comment]
Thanks so much. That's awesome!
I am able to fully reproduce all the LM results.

**Comment by DavidHerel** [comment]
Thanks so much. That's awesome!
I am able to fully reproduce all the LM results.
